include/watchdog.h

```diff
@@ -1,116 +1,206 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
+/**
+ * @brief Enable calling application callback function on watchdog timeout.
+ */
+#define WDT_MODE_CALLBACK	BIT(0)
 
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+
+/**
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs reloading in specified time window, otherwise
+ * the watchdog will trigger.
+ *
+ * @param up Upper limit of watchdog reload timeout in miliseconds.
+ * @param down Lower limit of watchdog reload timeout in miliseconds.
+ */
+struct wdt_window {
+	u16_t up;
+	u16_t down;
 };
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog configuration struct.
+ *
+ * @param options - a bitfield with following parts:
+ *    enable callback     [ 0 ]       - reset or callback -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
+struct wdt_cfg {
+	u8_t options;
```

Comment:
Sorry, I don't get this one.  Why is this a struct without named fields?  If you want a simple way to pass three flags, then just pass a flags word directly and not a pointer to a configuration struct.   If you're going to go through the trouble to make a struct, we should at least get typesafety out of the deal with e.g. single-bit fields named "callback_en", "pause_in_sleep" and "pause_in_debug" or whatever.

Reply:
I am aware of that, I will improve it in next iteration.